### PR TITLE
test: add tests for checkIfLegal and checkMove

### PR DESCRIPTION
**Pull request by eighty-three**

### CODE REVIEW

File: api/src/game/utils.spec.ts [hunk 1]
```diff
@@ -54,128 +71,53 @@ describe('testing checkMove', () => {
...
     });
 
-    test('to be 2, attacker is spy, target is private', () => {
-      expect(checkMove(gameStates, roomName, 'p1', 'A3', 'B3')).toBe(2);
+    test('it returns 2 when attacker is spy, target is a private', () => {
+      expect(checkMove(gameStates, roomName, 'p1', 'A5', 'A6')).toEqual(2);
+    });
+
+    test('it returns 2, attacker is spy, target is not a private', () => {
+      expect(checkMove(gameStates, roomName, 'p1', 'A5', 'A6')).toEqual(2);
```
Comment (eighty-three) [inline api/src/game/utils.spec.ts:103]: Repeating

File: api/src/game/utils.spec.ts [hunk 2]
```diff
@@ -54,128 +71,53 @@ describe('testing checkMove', () => {
...
     test('to be 3, result draw', () => {
-      expect(checkMove(gameStates, roomName, 'p1', 'A1', 'B1')).toBe(3);
+      expect(checkMove(gameStates, roomName, 'p1', 'H5', 'H6')).toEqual(3);
     });
 
-    test('to be 2, d > o', () => {
+    test('edge case: it returns 2 when origin and destination are at either sides of the board.', () => {
       /* Example of impossible move in the client but valid in server.
        * Only for testing purposes, it should normally be impossible
        */
```
Comment (eighty-three) [inline api/src/game/utils.spec.ts]: Changing the title of the test isn't necessary. If you want to specifically at this 'edge case' then you should've added a different test because now, 'to be 2, where d > o' isn't tested anymore

File: api/src/game/utils.spec.ts [hunk 3]
```diff
@@ -54,128 +71,53 @@ describe('testing checkMove', () => {
...
 
     test('to be 1, o > d', () => {
-      expect(checkMove(gameStates, roomName, 'p1', 'A4', 'B4')).toBe(1);
+      expect(checkMove(gameStates, roomName, 'p1', 'B5', 'B6')).toEqual(1);
     });
 
-    test('to be 1, o > d, empty space', () => {
-      expect(checkMove(gameStates, roomName, 'p1', 'A5', 'C1')).toBe(1);
+    test('it returns 1 when destination is an empty space', () => {
+      expect(checkMove(gameStates, roomName, 'p1', 'D3', 'D5')).toEqual(1);
```
Comment (eighty-three) [inline api/src/game/utils.spec.ts]: This was a mistake by the way, this comment:
```
      /* Example of impossible move in the client but valid in server.
       * Only for testing purposes, it should normally be impossible
       */
```
Should be for this test block.

File: api/src/game/utils.spec.ts [hunk 4]
```diff
@@ -54,128 +71,53 @@ describe('testing checkMove', () => {
...
-      p2: {
-        time: 1, start: true, name: 'p2',
-        board: {...board2}
-      }
-    };
 
-    // re side effects, see utils.ts lines 197-222
+    test('to be 1, attacker is private, target is not a spy', () => {
+      expect(checkMove(gameStates, roomName, 'p1', 'I5', 'F5')).toEqual(2);
+    });
```
Comment (eighty-three) [inline api/src/game/utils.spec.ts]: This is confusing because you're using the same coordinate as the origin for two different tests. It only works because of the `beforeEach` call you were doing, I think.

File: api/src/game/utils.spec.ts [hunk 5]
```diff
@@ -27,18 +27,35 @@ describe('testing checkIfLegal', () => {
...
       A1: { name: 'x', value: 1 },
-      A2: { name: 'x', value: 1 }
+      A2: { name: 'x', value: 1 },
+      B3: { name: 'x', value: 5 },
+      B5: { name: 'x', value: 5 },
+      A5: { name: 'x', value: 99 },
+      C2: { name: 'x', value: 5 },
+      D3: { name: 'x', value: 5 },
+      H5: { name: 'x', value: 2 },
+      I5: { name: 'x', value: 2 }
```
Comment (eighty-three) [inline api/src/game/utils.spec.ts]: If you really think spreading out the pieces is necessary you should at least sort them in the tests
Functionally there's no difference between these two objects
```
{ A1, B1, C1, D1, E1, F1 }

{ A1, E1, F1, C1, B1, D1 }
```
The latter's purpose is only to make things confusing.

**Review by eighty-three — COMMENTED** [review]
I don't think it's necessary to spread out the pieces because it doesn't make a difference when and where the pieces move. You don't need to simulate the board because the function should work in a vacuum. The previous tests were made not to make the function work (retrofitting, as you've mentioned) but to make reading the tests easier.

**Comment by eighty-three** [comment]
If I want to test these three things:
1. o > d
2. d > o
3. o === d

Does it mean that because I tested it all from A1 to B1, A2 to B2, A3 to B3, the tests are invalid because it's not tested from
C1 to C2, B5 to A5, G7 to H7?
There's literally no difference, we are testing the same function.

It makes a difference for the 3rd describe block, the `side effects` because the positions are relevant (for the flag, for example)
but for the specific functionality, the movement of the piece/vs.ing vs opponent pieces, it doesn't matter where or what position it's tested at.

File: api/src/game/utils.spec.ts (hunk 4 above)
Comment (techiesspammer) [reply api/src/game/utils.spec.ts]: Yes, that's right. Since we are calling `BeforeEach` on each `test` block, the board gets reset.

File: api/src/game/utils.spec.ts (hunk 2 above)
Comment (techiesspammer) [reply api/src/game/utils.spec.ts]: Okay. I'll change it back

File: api/src/game/utils.spec.ts (hunk 3 above)
Comment (techiesspammer) [reply api/src/game/utils.spec.ts]: This comment is applicable for the edge case test though..
